WP_Theme_JSON: deprecate to_ruleset() and replace with Style Engine compile_css()

ramonjd opened the pull request:
## Context

Part of https://github.com/WordPress/gutenberg/issues/65728

This will be a series of small, and probably very infrequent PRs to move styles-related logic from `WP_Theme_JSON` to the style engine. 

There's a [rough plan over at the issue](https://github.com/WordPress/gutenberg/issues/65728#issuecomment-3507290654). The strategy is to knock things off one by one with the end goal being that the [style engine can generated CSS from a theme.json tree](https://github.com/WordPress/gutenberg/issues/65728#issuecomment-2400888663).

## What

Deprecates `WP_Theme_JSON_Gutenberg::to_ruleset()` and replaces all internal calls with `WP_Style_Engine_Gutenberg::compile_css()`. 

## Why

This is the first phase of extracting style/CSS logic from `WP_Theme_JSON` into the Style Engine (see [#65728](https://github.com/WordPress/gutenberg/issues/65728)). The `to_ruleset()` method duplicates functionality already present in the Style Engine, and consolidating CSS generation improves maintainability and reduces code duplication.

## How

1. **Added private methods `normalize_declarations()` and `convert_theme_json_declarations()` to handle format conversion between theme.json's array-of-arrays format and the Style Engine's associative array format.

2. **Enhanced `WP_Style_Engine::compile_css()`** to auto-detect and convert theme.json format declarations using the new utility class.

3. **Replaced all 9 internal calls** to `to_ruleset()` with `WP_Style_Engine_Gutenberg::compile_css()` throughout `WP_Theme_JSON_Gutenberg`.

4. **Deprecated `to_ruleset()` method** with `@deprecated` tag and `_deprecated_function()` call, keeping it as a backwards-compatible wrapper that delegates to the Style Engine.

5. **Updated test expectations** to match Style Engine's CSS output format (no spaces after colons in declarations).

## Testing instructions

1. **Run PHP unit tests:**
 
- `npm run test:unit:php:base -- --filter WP_Theme_JSON_Gutenberg_Test`
- `npm run test:unit:php:base -- --filter WP_Style_Engine`

3. **Verify deprecation notice:**
- The `to_ruleset()` method should trigger a deprecation notice if called directly (though all internal calls have been replaced).
- Check that the deprecation message points to `WP_Style_Engine_Gutenberg::compile_css()` as the alternative.

4. **Visual regression:**
- Test theme.json styles render correctly in the editor and frontend.
- Verify preset classes (`.has-*-color`, `.has-*-font-size`, etc.) work as expected.
- Check layout styles, duotone filters, and style variations render properly.

ramonjd committed:
Refactor WP Style Engine: Introduce WP_Style_Engine_Utils for theme.json compatibility

- Updated `compile_css` method to utilize `WP_Style_Engine_Utils` for normalizing CSS declarations from theme.json format.
- Added new utility class `WP_Style_Engine_Utils` with methods for converting theme.json declarations to associative array format.
- Included tests for the new utility functions to ensure proper conversion and handling of various input formats.
- Updated relevant files to reflect changes in CSS compilation logic.

This refactor enhances the flexibility of the style engine by supporting both associative array and theme.json formats for CSS declarations.

ramonjd committed:
Refactor CSS Compilation in WP_Theme_JSON_Gutenberg

- Replaced static::to_ruleset calls with WP_Style_Engine_Gutenberg::compile_css for improved CSS generation.
- Updated various sections to ensure consistent handling of CSS declarations and selectors.
- Adjusted test cases to reflect changes in CSS output formatting.

This refactor enhances the maintainability and performance of the CSS compilation process within the theme JSON handling.

ramonjd committed:
Deprecate to_ruleset Method in WP_Theme_JSON_Gutenberg

- Marked the static::to_ruleset method as deprecated in favor of WP_Style_Engine_Gutenberg::compile_css for improved CSS handling.
- Updated the method to delegate CSS compilation to the new style engine, ensuring better format detection and conversion.

This change enhances the maintainability of the code by consolidating CSS generation logic.

ramonjd commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -28,7 +28,7 @@ class WP_Theme_JSON_Gutenberg_Test extends WP_UnitTestCase {
 	 *
```
All the changes in this file are whitespace changes between property and value in CSS declarations.

The Style Engine outputs rules without whitespace, e.g., `font-size:60px;`, whereas the theme json class outputs `font-size: 60px;`

ramonjd commented on packages/style-engine/src/class-wp-style-engine.php:
```diff
@@ -639,8 +639,13 @@ protected static function get_url_or_value_css_declaration( $style_value, $style
 		 *
-		 * @param string[] $css_declarations An associative array of CSS definitions, e.g., array( "$property" => "$value", "$property" => "$value" ).
-		 * @param string   $css_selector     When a selector is passed, the function will return a full CSS rule `$selector { ...rules }`, otherwise a concatenated string of properties and values.
+		 * Accepts declarations in either format:
```
```suggestion
		 * For backwards compatibility, accepts declarations in two formats:
```

ramonjd commented on packages/style-engine/src/class-wp-style-engine.php:
```diff
@@ -639,8 +639,13 @@ protected static function get_url_or_value_css_declaration( $style_value, $style
-		 * @param string   $css_selector     When a selector is passed, the function will return a full CSS rule `$selector { ...rules }`, otherwise a concatenated string of properties and values.
+		 * Accepts declarations in either format:
+		 * - Associative array: array( 'property' => 'value', ... )
+		 * - Theme.json format: array( array( 'name' => 'property', 'value' => 'value' ), ... )
```
```suggestion
		 * - Theme.json format (deprecated): array( array( 'name' => 'property', 'value' => 'value' ), ... )
```

ramonjd commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -1949,24 +1949,17 @@ protected function get_css_variables( $nodes, $origins ) {
-		if ( empty( $declarations ) ) {
-			return '';
-		}
+		_deprecated_function( __METHOD__, 'Gutenberg 7.0.0', 'WP_Style_Engine_Gutenberg::compile_css()' );
```
In Core, this will be `WP_Style_Engine::compile_css()`

andrewserong replied:
This is a protected method, does it need a `_deprecated_function` call, or can it just be removed? I don't see any calls to it in `wordpress-develop` other than internal calls within the Theme JSON class.

Or is it to protect classes (like this one in Gutenberg) that might be extending the Theme JSON class? If so, then yes, deprecating sounds safer! (Did I just answer my own question? 😆)

ramonjd replied:
It's a good question. I'd actually be happy to delete it.

But I did it for the reason you mention (in case it's been extended) and this is despite the 4yo notice at the top of the class that says

```php
/**
 * Class that encapsulates the processing of structures that adhere to the theme.json spec.
 *
 * This class is for internal core usage and is not supposed to be used by extenders (plugins and/or themes).
 * This is a low-level API that may need to do breaking changes. Please,
 * use get_global_settings, get_global_styles, and get_global_stylesheet instead.
 *
 * @access private
 */
```

So do we care about folks who extended anyway? 🤔

youknowriad commented on packages/style-engine/src/class-wp-style-engine-utils.php:
```diff
@@ -0,0 +1,76 @@
+		/**
+		 * Normalizes CSS declarations to associative array format.
+		 *
+		 * Detects and converts theme.json format (array of arrays with 'name' and 'value' keys)
```
Where does this "source" format comes from, Is this some kind of internal format?

ramonjd replied:
The `$declarations` array the Theme JSON `to_ruleset()` method accepts has always been:

```
Input: array(array('name' => 'property', 'value' => 'value'), ...)
Output: "selector {property: value;}"
```

Is that what you're referring to as "source"? Theme JSON has always built CSS declarations according to this model.

The Style Engine uses associative arrays:

```
Input: array('property' => 'value', ...) (associative array)
Output: "selector {property: value;}" 
```

So this conversion is a temporary measure to make the style engine compatible with with Theme JSON style declaration objects.

It's "temporary" because, over time, as [more pieces are moved away](https://github.com/WordPress/gutenberg/issues/65728#issuecomment-3507290654) from the Theme JSON class the `name/value` format will be redundant. The style engine Theme JSON logic will handle the building of style nodes, then declarations, and the creation of CSS from them. 

That's the working theory. I tried a few things and the changes would have been too radical and prone to bugs, so I was thinking of a slower, more methodical approach to tame this beast. 

Alternatively we could rewire the style engine to exclusively use `array(array('name' => 'property', 'value' => 'value'), ...)`. We made the style engine class `final` so we could change things like this.

I'm not sure how much value that would being, aside from conformity. Theme JSON methods that currently generate and return declarations are all protected.

youknowriad replied:
I see, wouldn't we be creating public APIs if we ship this class which means it won't be temporary? Maybe private methods would work better?

ramonjd replied:
Good point. I concur! I'll do a switcheroo to private 🙇🏻

ramonjd committed:
Apply suggestions from code review

ramonjd committed:
Refactor WP Style Engine: Remove WP_Style_Engine_Utils and Integrate Functionality

- Deleted the WP_Style_Engine_Utils class and its associated methods for normalizing CSS declarations.
- Integrated the normalization logic directly into the WP_Style_Engine class, enhancing encapsulation and reducing dependencies.
- Updated references in the load.php and package.json files to reflect the removal of the utility class.
- Adjusted tests to ensure compatibility with the new structure.

This refactor streamlines the codebase by consolidating functionality and improving maintainability.

ramonjd committed:
Add new props to safe attrs, fix tests, ensure that style engine can handle simple number values

ramonjd committed:
Fix margin-block-start: null handling

ramonjd commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -1812,10 +1812,15 @@ protected function get_layout_styles( $block_metadata, $options = array() ) {
-								foreach ( $spacing_rule['rules'] as $css_property => $css_value ) {
-									$current_css_value = is_string( $css_value ) ? $css_value : $block_gap_value;
-									if ( static::is_safe_css_declaration( $css_property, $current_css_value ) ) {
+							// Iterate over each of the styling rules and substitute non-string values such as `null` with the real `blockGap` value.
```
Trunk uses `to_ruleset` — which builds CSS via string concatenation `(`$dec['name'] . ': ' . $dec['value'])`. 

When `$dec['value']` is `bool(true)`, PHP's implicit coercion converts it to '1' and produces `margin-block-start: 1`. That's why trunk tests pass without any fix.
